packages/providers/src/provider.test.ts

```diff
@@ -244,4 +251,343 @@ describe('Provider', () => {
+    });
+
+    try {
+      await provider.sendTransaction(transactionRequest);
+    } catch (e) {
+      const EXCLUDED = provider.cache?.getExcluded() || [];
+      expect(EXCLUDED.length).toEqual(3);
+      expect(EXCLUDED.map((value) => hexlify(value))).toStrictEqual(EXPECTED);
+
+      // clear cache
+      EXCLUDED.forEach((value) => provider.cache?.del(value));
+    }
+  });
+
+  it('can cacheUtxo [will cache inputs and also merge/de-dupe in exclude list]', async () => {
+    const provider = new Provider('http://127.0.0.1:4000/graphql', { cacheUtxo: 10000 });
+    const EXPECTED: BytesLike[] = [
+      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c503',
+      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c504',
+      '0xda5d131c490db3868be9f8e228cf279bd98ef1de97129682777ed93fa088bc3505',
+    ];
+    const MessageInput: MessageTransactionRequestInput = {
+      type: InputType.Message,
+      amount: 100,
+      sender: NativeAssetId,
+      recipient: NativeAssetId,
+      witnessIndex: 1,
+      data: NativeAssetId,
+      nonce: 1,
+    };
+    const CoinInputA: CoinTransactionRequestInput = {
+      type: InputType.Coin,
+      id: EXPECTED[0],
+      owner: NativeAssetId,
+      assetId: NativeAssetId,
+      txPointer: NativeAssetId,
+      witnessIndex: 1,
+      amount: 100,
+    };
+    const CoinInputB: CoinTransactionRequestInput = {
+      type: InputType.Coin,
+      id: arrayify(EXPECTED[1]),
+      owner: NativeAssetId,
+      assetId: NativeAssetId,
+      txPointer: NativeAssetId,
+      witnessIndex: 1,
+      amount: 100,
+    };
+    const CoinInputC: CoinTransactionRequestInput = {
+      type: InputType.Coin,
+      id: EXPECTED[2],
+      owner: NativeAssetId,
+      assetId: NativeAssetId,
+      txPointer: NativeAssetId,
+      witnessIndex: 1,
+      amount: 100,
+    };
+    const transactionRequest = new ScriptTransactionRequest({
+      inputs: [MessageInput, CoinInputA, CoinInputB, CoinInputC],
+    });
+
+    try {
+      await provider.sendTransaction(transactionRequest);
+    } catch (e) {
+      const EXCLUDED = provider.cache?.getExcluded() || [];
+      expect(EXCLUDED.length).toEqual(3);
+      expect(EXCLUDED.map((value) => hexlify(value))).toStrictEqual(EXPECTED);
+
+      const owner = Address.fromRandom();
+      const resourcesToSpendMock = jest.fn(() => Promise.resolve({ resourcesToSpend: [] }));
+      // @ts-expect-error mock
+      provider.operations.getResourcesToSpend = resourcesToSpendMock;
+      await provider.getResourcesToSpend(owner, [], {
+        utxos: [
+          '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c503',
+          '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c507',
+          '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c508',
+        ],
+      });
+
+      expect(resourcesToSpendMock).toHaveBeenCalledWith({
```

Comment:
> **Warning** This must be reviewed for all tests.

Several tests developed a pattern of having expectations <sup>`expect(...)` calls</sup> inside `catch` blocks. This doesn't work because `expect()`-ations must always run, while `catch` blocks might **_not_** — for instance, if the code inside the `try` block doesn't throw. This can result in false positives.

You can circumvent this by placing them outside the `catch` block.

```ts
let error

try {
  // will always run
  doSomething();
} catch (err) {
  // might not run
  error = err;
}

expect(error).toBeTruthy(); // ensures method threw (or not)
expect(...); // other validations
```

Reply:
@camsjams I wrote a simple test utility for this on the Typegen package. I can quickly export it from a shared package if you think it'd be worth using it on this PR — just let me know.

<details>
<summary>Utility</summary>

https://github.com/FuelLabs/fuels-ts/blob/9d70dfa4f2bcef2caeefac9cd358176c43a4b59a/packages/abi-typegen/test/utils/executeAndCatch.ts#L1-L12
</details>


<details>
<summary>Usage 1</summary>

https://github.com/FuelLabs/fuels-ts/blob/9d70dfa4f2bcef2caeefac9cd358176c43a4b59a/packages/abi-typegen/src/runTypegen.test.ts#L196-L209
</details>


<details>
<summary>Usage 2</summary>

https://github.com/FuelLabs/fuels-ts/blob/9d70dfa4f2bcef2caeefac9cd358176c43a4b59a/packages/abi-typegen/src/utils/findType.test.ts#L34-L38
</details>

Reply:
Cool, will do.

Reply:
Here:
 - #978